closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java

```diff
@@ -0,0 +1,1040 @@
+                                + " Algorithm " + algorithm
+                                + " is allowed in provider " + providerName);
+                    }
+                    return true;
+                }
+
+                // For type and algorithm match, and attribute is not *.
+                // Then continue checking attributes.
+                if (cTypePut && cAlgorithmPut) {
+                    String[] cAttributeArray = constraint.attributes.split(":");
+
+                    // For each attribute, must be all matched for return allowed.
+                    for (String attribute : cAttributeArray) {
+                        String[] input = attribute.split("=", 2);
+
+                        String cName = input[0].trim();
+                        String cValue = input[1].trim();
+                        String sValue = service.getAttribute(cName);
+                        if ((sValue == null) && !cValue.equalsIgnoreCase(sValue)) {
+                            // Any of the attribute not match, return service is not allowed
+                            return false;
+                        }
+                    }
+                    if (debug != null) {
+                        debug.println(
+                                "Security constraints check."
+                                        + " Service type: " + type
+                                        + " Algorithm: " + algorithm
+                                        + " Attribute: " + constraint.attributes
+                                        + " is allowed in provider: " + providerName);
+                    }
+                    return true;
+                }
+            }
+            if (debug != null) {
+                debug.println("Security constraints check."
+                        + " Service type: " + type
+                        + " Algorithm: " + algorithm
+                        + " is NOT allowed in provider " + providerName);
+            }
+            // No match for any constraint, return NOT allowed.
+            return false;
+        }
+
+        /**
+         * Check if the provider is allowed in restricted security mode.
+         *
+         * @param providerName the provider to check
+         * @return true if the provider is allowed
+         */
+        boolean isRestrictedProviderAllowed(String providerName) {
+            if (debug != null) {
+                debug.println("Checking the provider " + providerName + " in restricted security mode.");
+            }
+
+            // Remove the provider class package name if there is.
+            int pos = providerName.lastIndexOf('.');
+            providerName = (pos < 0) ? providerName : providerName.substring(pos + 1, providerName.length());
+
+            // Remove argument, e.g. -NSS-FIPS, if there is.
+            pos = providerName.indexOf('-');
+            providerName = (pos < 0) ? providerName : providerName.substring(0, pos);
```

Comment:
Perhaps this substring extraction should happen in the reverse order (select the portion before a `-`, _then_ choose the part after the last `.`)?

Reply:
I don't see a response to this question.

Reply:
@keithc-ca 

These codes, first remove the class package name if there is, which is in the fix format using ‘.’ to separate the packages. So, after remove the package name, we can make sure only the class name left. And then remove the argument(e.g. -NSS-FIPS) from the class name if there is. 

If reverse the order, first to remove the class name argument, which start with ‘-‘, and then remove the class package name. But if there is a ‘-‘ also in the package name, for example, com.test-provider.SunPKCS11-NSS-FIPS. If I first select the portion before a ‘-‘, then the “com.test” will be returned and it is wrong.

So I suggest, we first choose the part after the last ‘.’ which is the fix format for the class packages, to make sure we get the correct class name, and then select the portion before a ‘-‘ to remove the class name argument if there is.

Reply:
The example you give (`com.test-provider.SunPKCS11-NSS-FIPS`) is not possible to write in the Java (source) language. Yes, the JVM will accept classes and package names containing `-`, but you have to write them in some other language (and you can't use them directly in Java source code).

I was thinking of a different example, say `com.sample.provider.SunPKCS11-NSS-FIPS.v2`. The current code will select `v2` which is probably not what we want.

How would you propose we deal with these (exceptional) situations?